src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -0,0 +1,137 @@
+        """
+        # TODO: message can contain filters
+        data = {"event": "offers_bids_response"}
+        market_offers_bids_list_mapping = {}
+        for area_uuid, markets in self.area_uuid_markets_mapping.items():
+            for market in markets:
+                self.markets_mapping[market.id] = market
+                market_offers_bids_list_mapping[market.id] = {"bids": [], "offers": []}
+                bids, offers = market.open_bids_and_offers
+                market_offers_bids_list_mapping[market.id]["bids"].extend(
+                    list(bid.serializable_dict() for bid in bids.values()))
+                market_offers_bids_list_mapping[market.id]["offers"].extend(
+                    list(offer.serializable_dict() for offer in offers.values()))
+        data.update({
+            "market_offers_bids_list_mapping": market_offers_bids_list_mapping,
+        })
+
+        channel = f"{self.channel_prefix}response/get_offers_bids/"
+        self.myco_ext_conn.publish_json(channel, data)
+
+    def match_recommendations(self, message):
+        """
+        Receive trade recommendations and match them in the relevant market
+        """
+        channel = f"{self.channel_prefix}response/matched_recommendations/"
+        response_dict = {"event": "match", "status": "success"}
+        data = json.loads(message.get("data"))
+        recommendations = data.get("recommended_matches", [])
+        validated_records = {}
+        for record in recommendations:
+            market = self.markets_mapping.get(record.get("market_id"), None)
+            if market is None or market.readonly:
+                # The market is already finished or doesn't exist
+                del record
+                continue
+
+            bid = bid_from_JSON_string(json.dumps(record.get("bid")))
+            offer = offer_from_JSON_string(json.dumps(record.get("offer")),
+                                           record.get("bid").get("time"))
+
+            try:
+                validate_authentic_bid_offer_pair(
+                    bid,
+                    offer,
+                    record.get("trade_rate"),
+                    record.get("selected_energy")
+                    )
+                assert offer.id in market.offers
+                assert bid.id in market.bids
+                if record.get("market_id") not in validated_records:
+                    validated_records[record.get("market_id")] = []
+
+                validated_records[record.get("market_id")].append(BidOfferMatch(
+                    bid,
+                    record.get("selected_energy"),
+                    offer,
+                    record.get("trade_rate")))
+            except AssertionError:
+                # If validation fails or offer/bid were consumed
+                response_dict["status"] = "fail"
+                response_dict["message"] = "Validation Error"
+                break
```

Comment:
I would highly recommend replacing assertions with custom exceptions. As a best practice, it's recommended to use assertions only to check for code issues that need to be spotted during the development/debug phase (for coding mistakes, for example). 
For cases in which the program can fail because of external input (like in this case) the best thing is to use exceptions and handle them.

One more reason to use exceptions (that doesn't apply to our current case, but might in the future) is that `assert` statements can be automatically removed if optimizations are requested by Python.

Reply:
Some sources that explain it clearly (in my opinion):
- https://softwareengineering.stackexchange.com/a/225996
- https://stackoverflow.com/questions/40182944/whats-the-difference-between-raise-try-and-assert

Reply:
(So usually you want assertions to fail badly and not to be catched, because they're meant to be seen and fixed by the developer. Exceptions are meant to be catched instead, since they are closely related to the production code.)